### PR TITLE
Add spec test support for V128.

### PR DESCRIPTION
**Pull request by nlewycky**

### CODE REVIEW

**Comment by binji** [comment]
Thanks for looking into this! Can you describe what the output looks like here? I'm having a bit of trouble telling from the code.

Also, for this to work, you need to parse these values too in `JSONParser::ParseConst` in `spectest-interp.cc`

**Comment by nlewycky** [comment]
The output is a plain decimal integer. I've added the parser.

**Comment by syrusakbary** [comment]
Some context on this PR: we are working on adding SIMD support to the Wasmer runtime.

Wasmer relies on wabt for parsing the `.wast` tests, and we are on the finish line to make all tests pass 🎉

Let us know if you have more feedback on what we can improve in the PR! :)

File: src/literal.cc [hunk 1]
```diff
@@ -622,6 +622,74 @@ Result ParseInt64(const char* s,
...
+  v->v[1] = add_with_carry(v->v[1] << 3, v->v[1] << 1, &carry_into_v2);
+  v->v[1] = add_with_carry(v->v[1], carry_into_v1, &carry_into_v2);
+  uint32_t carry_into_v3 = ((v->v[2] & 0xe0000000) >> 29) +
+                           ((v->v[2] & 0x80000000) >> 31);
+  v->v[2] = add_with_carry(v->v[2] << 3, v->v[2] << 1, &carry_into_v3);
+  v->v[2] = add_with_carry(v->v[2], carry_into_v2, &carry_into_v3);
+  v->v[3] = v->v[3] * 10 + carry_into_v3;
+}
+
+Result ParseUint128(const char* s,
```
Comment (binji) [inline src/literal.cc]: Please add some tests for this (see test-literal.cc), and also `WriteV128`

File: src/binary-writer-spec.cc [hunk 2]
```diff
@@ -200,6 +202,43 @@ void BinaryWriterSpec::WriteConst(const Const& const_) {
...
+      WriteString("v128");
+      WriteSeparator();
+      WriteKey("value");
+
+      v128 v = const_.v128_bits;
+
+      uint64_t digits, remainder;
+      std::ostringstream reversed_result;
+      do {
+        remainder = v.v[3];
```
Comment (binji) [inline src/binary-writer-spec.cc]: If you can, I'd prefer if this was rolled up into a 4 element loop instead.

File: src/literal.cc [hunk 3]
```diff
@@ -622,6 +622,74 @@ Result ParseInt64(const char* s,
...
+}
+
+static void mul_10(v128* v) {
+  // Multiply-by-10 decomposes into (x << 3) + (x << 1). We implement those
+  // operations with carrying from smaller quads of the v128 to the larger
+  // quads.
+
+  // 0xe0_00_00_00 is a mask for the top 3 bits, to carry over from x << 3.
+  // 0x80_00_00_00 does the same for x << 1.
+  uint32_t carry_into_v1 = ((v->v[0] & 0xe0000000) >> 29) +
```
Comment (binji) [inline src/literal.cc]: nit: Use a `constexpr` variable for `0xe0000000` and `0x80000000`

File: src/binary-writer-spec.cc [hunk 4]
```diff
@@ -200,6 +202,43 @@ void BinaryWriterSpec::WriteConst(const Const& const_) {
...
     }
 
+    case Type::V128: {
+      WriteString("v128");
+      WriteSeparator();
+      WriteKey("value");
+
+      v128 v = const_.v128_bits;
+
+      uint64_t digits, remainder;
```
[the remark this answers is not in the document]
Comment (nlewycky) [reply src/binary-writer-spec.cc]: Done.

[the remark this answers is not in the document]
Comment (nlewycky) [reply src/binary-writer-spec.cc]: Done.

File: src/binary-writer-spec.cc [hunk 5]
```diff
@@ -200,6 +202,43 @@ void BinaryWriterSpec::WriteConst(const Const& const_) {
...
+      std::ostringstream reversed_result;
```
[the remark this answers is not in the document]
Comment (nlewycky) [reply src/binary-writer-spec.cc]: Done.

File: src/literal.cc [hunk 6]
```diff
@@ -622,6 +622,74 @@ Result ParseInt64(const char* s,
   return result;
 }
 
+static uint32_t add_with_carry(uint32_t x, uint32_t y, uint32_t* carry) {
```
[the remark this answers is not in the document]
Comment (nlewycky) [reply src/literal.cc]: Done.

[the remark this answers is not in the document]
Comment (nlewycky) [reply src/literal.cc]: Done.

File: src/literal.cc (hunk 1 above)
Comment (nlewycky) [reply src/literal.cc]: Done.

File: src/binary-writer-spec.cc (hunk 2 above)
Comment (nlewycky) [reply src/binary-writer-spec.cc]: Rolled three into a loop but not four.

File: src/literal.cc (hunk 3 above)
Comment (nlewycky) [reply src/literal.cc]: Done.

**Pull request merged**